src/sdk/namespace_fs.js

```diff
@@ -1964,70 +1980,43 @@ class NamespaceFS {
-                }
-            } else {
-                const now = new Date();
-
-                // The only entity that can remove the restore request is the underlying
-                // restore mechanism, implying that at some point it must have picked
-                // up the request to restore and finished the process.
-                if (restore_expiry && restore_expiry > now) {
-                    dbg.log0('namespace_fs.restore_object: state - (!restore_request, !restore_ongoing, restore_expiry > now)');
+            const now = new Date();
+            const restore_state = this._get_restore_status(stat.xattr, now, file_path);
+            dbg.log1(
+                'namespace_fs.restore_object:', file_path,
+                'restore_state:', restore_state,
+            );
 
-                    const expires_on = now;
-                    expires_on.setDate(expires_on.getDate() + params.days);
+            if (restore_state.state === RESTORE_STATUS_CAN_RESTORE) {
+                // First add it to the log and then add the extended attribute as if we fail after
+                // this point then the restore request can be triggered again without issue but
+                // the reverse doesn't works.
+                await this.append_to_restore_wal(file_path);
 
-                    await file.replacexattr(fs_context, {
-                        [GlacierBackend.XATTR_RESTORE_EXPIRY]: expires_on.toISOString(),
-                    });
+                await file.replacexattr(fs_context, {
+                    [GlacierBackend.XATTR_RESTORE_REQUEST]: params.days.toString(),
+                });
 
-                    // Should result in HTTP: 200 OK
-                    return false;
-                }
+                // Should result in HTTP: 202 Accepted
+                return true;
+            }
 
-                // If neither the restore is ongoing nor a restore expiry is set, we can set the xattr
-                // essentially adding the object to the batch.
-                if (!restore_expiry || restore_expiry <= now) {
-                    dbg.log0('namespace_fs.restore_object: state - (!restore_request, !restore_ongoing, !restore_expiry or restore_expiry <= now)');
+            if (restore_state.state === RESTORE_STATUS_ONGOING) {
+                throw new S3Error(S3Error.RestoreAlreadyInProgress);
+            }
 
-                    // First add it to the log and then add the extended attribute as if we fail after
-                    // this point then the restore request can be triggered again without issue but
-                    // the reverse doesn't works.
-                    await this.append_to_restore_wal(file_path);
+            if (restore_state.state === RESTORE_STATUS_RESTORED) {
+                const expires_on = now;
+                expires_on.setUTCDate(expires_on.getUTCDate() + params.days);
+                expires_on.setUTCHours(0, 0, 0, 0);
```

Comment:
Should we really use midnight in UTC instead of midnight in local TZ?
Perhaps we could have the expiry time-of-day as a config.

Reply:
This was intentional so as to not rely on all the nodes (nsfs nodes) having the same TZ configured. Do we need the TZ info?